packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts

```diff
@@ -103,18 +103,41 @@ export async function uploadArtifactToS3(
     try {
         const uploadFileByteSize = (await nodefs.promises.stat(fileName)).size
         getLogger().info(
-            `Uploading project artifact at %s with checksum %s using uploadId: %s and size %s kB`,
+            `CodeTransformation: Uploading project artifact at %s with checksum %s using uploadId: %s and size %s kB`,
             fileName,
             sha256,
             resp.uploadId,
             Math.round(uploadFileByteSize / 1000)
         )
 
-        const response = await request.fetch('PUT', resp.uploadUrl, {
-            body: buffer,
-            headers: getHeadersObj(sha256, resp.kmsKeyArn),
-        }).response
-        getLogger().info(`CodeTransformation: Status from S3 Upload = ${response.status}`)
+        let response = undefined
+        const retriableCodes = [408, 429, 500, 502, 503, 504]
+        for (let i = 0; i < 3; i++) {
+            try {
+                response = await request.fetch('PUT', resp.uploadUrl, {
```

Comment:
(Non-blocker) Looks like `waitUntil` wouldn't work here because it only has "max time", not "max retries". Probably should add that. Being be able to "find all waitUntil references" can be useful to see all the areas in the codebase that are doing this pattern.

Reply:
On second thought, I don't see a point in retrying the upload if we know it failed for non-retriable reasons, so I will keep the PR as it is for now. Updated the comment to reflect this.

Reply:
> But it looks like it will always retry until the timeout period elapses, compared to this function which will only retry as long as the status code is retriable

Your callback passed to `waitUntil` can decide whether to retry by its return value. However, it would be nicer if waitUntil supported a more explicit way to stop retrying.